audio/src/passthrough_decoder.rs

```diff
@@ -0,0 +1,195 @@
+        info!("Starting passthrough track with serial {}", stream_serial);
+
+        return Ok(PassthroughDecoder {
+            rdr,
+            wtr,
+            lastgp_page: Some(0),
+            absgp_page: 0,
+            stream_serial,
+        });
+    }
+}
+
+impl<R: Read + Seek> AudioDecoder for PassthroughDecoder<R> {
+    fn seek(&mut self, ms: i64) -> Result<(), AudioError> {
+        info!("Seeking to {}", ms);
+        self.lastgp_page = match ms {
+            0 => Some(0),
+            _ => None,
+        };
+
+        // hard-coded to 44.1 kHz
+        match self.rdr.seek_absgp(None, (ms * 44100 / 1000) as u64) {
+            Ok(_) => return Ok(()),
+            Err(err) => return Err(AudioError::PassthroughError(err.into())),
+        }
+    }
+
+    fn next_packet(&mut self) -> Result<Option<AudioPacket>, AudioError> {
+        let mut skip = self.lastgp_page.is_none();
+        loop {
+            let pck = match self.rdr.read_packet() {
+                Ok(Some(pck)) => pck,
+
+                Ok(None) | Err(OggReadError::NoCapturePatternFound) => {
+                    info!("end of streaming");
+                    return Ok(None);
+                }
+
+                Err(err) => return Err(AudioError::PassthroughError(err.into())),
+            };
+
+            let pckgp_page = pck.absgp_page();
+            let lastgp_page = self.lastgp_page.get_or_insert(pckgp_page);
+
+            // consume packets till next page to get a granule reference
+            if skip {
+                if *lastgp_page == pckgp_page {
+                    debug!("skipping packet");
+                    continue;
+                }
+                skip = false;
+                info!("skipped at {}", pckgp_page);
+            }
+
+            // now we can calculate absolute granule
+            self.absgp_page += pckgp_page - *lastgp_page;
+            self.lastgp_page = Some(pckgp_page);
+
+            // set packet type
+            let inf = if pck.last_in_stream() {
+                self.lastgp_page = Some(0);
+                PacketWriteEndInfo::EndStream
+            } else if pck.last_in_page() {
+                PacketWriteEndInfo::EndPage
+            } else {
+                PacketWriteEndInfo::NormalPacket
+            };
+
+            if let Err(err) = self.wtr.write_packet(
+                pck.data.into_boxed_slice(),
+                self.stream_serial,
+                inf,
+                self.absgp_page,
+            ) {
+                return Err(AudioError::PassthroughError(err.into()));
+            }
+
+            let data = self.wtr.inner_mut();
+
+            if data.len() > 0 {
+                let result = AudioPacket::OggData(std::mem::take(data));
+                return Ok(Some(result));
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum PassthroughError {
+    NotVorbisError,
+    OggError(OggReadError),
+    IOError(io::Error),
+}
```

Comment:
I think the `IOError` case will never occur, but I won't guarantee it: This variant is used only if `PacketWriter<Vec<u8>>::write_packet` fails. Looking at the source code, it seems as if this function only passes through the errors of the underlying `Write` object. But writing to a `Vec` never fails, so you could as well `unwrap` the `Result`s of `write_packet`. (Is there anyone who could confirm this?)

If this variant is removed, you could also remove the `NotVorbisError` variant and replace it by `OggReadError::InvalidData`. So there wouldn't be a need for an enum anymore, you could simply use a "newtype" struct as in the vorbis decoders.

Reply:
yes, more simplification :-)